### PR TITLE
feat: Switch Adaptation Project's generator/wizard to use Open Source writers

### PR DESCRIPTION
**Pull request by IvoSG**
Feat for https://github.com/SAP/open-ux-tools/issues/1611.

Adds writer functionality for adaptation projects.

### CODE REVIEW

File: packages/adp-tooling/test/unit/writer/__snapshots__/index.test.ts.snap [hunk 1]
```diff
@@ -63,18 +62,17 @@ Object {
...
     \\"keywords\\": [
         \\"ui5\\",
         \\"sapui5\\",
         \\"adaptation-project\\"
     ],
-    \\"dependencies\\": {},
     \\"devDependencies\\": {
         \\"@sap/ux-ui5-tooling\\": \\"^1.11.4\\",
-        \\"@ui5/cli\\": \\"^3.9.1\\"
+        \\"@ui5/cli\\": \\"^3.6.0\\"
```
Comment (nikmace) [inline packages/adp-tooling/test/unit/writer/__snapshots__/index.test.ts.snap:75]: Just a question, why did this version change? It was `^3.9.1` before. `^3.6.0` will install the latest too.

File: .changeset/famous-panthers-occur.md [hunk 2]
```diff
@@ -0,0 +1,6 @@
+---
+'@sap-ux/adp-tooling': minor
+'@sap-ux/preview-middleware': patch
+---
+
+Adds writer functionality for generation of Adaptation Project's
```
Comment (nikmace) [inline .changeset/famous-panthers-occur.md:6]: Spelling -> Projects

File: packages/adp-tooling/src/types.ts [hunk 3]
```diff
@@ -347,3 +364,49 @@ export interface AdpProjectData {
     reference: string;
     id: string;
 }
+
+export interface AdpCustomConfig {
+    adp?: {
+        safeMode: boolean;
```
Comment (nikmace) [inline packages/adp-tooling/src/types.ts:370]: Should we also have an `environment?: 'ABAP' | 'S4HANA' | 'CF'` under `adp`?

**Review by nikmace — CHANGES_REQUESTED** [review]
Coverage is currently at 88.7% and could be improved a little bit.

**Review by nikmace — CHANGES_REQUESTED** [review]

File: packages/adp-tooling/templates/projects/onprem/package.json [hunk 4]
```diff
@@ -14,7 +14,7 @@
         "@sap-ux/preview-middleware": "^0.11.1",
         "@sap-ux/ui5-proxy-middleware": "^1.3.0",
         "@sap-ux/deploy-tooling": "^0.11.7"<%}%>,
-        "@ui5/cli": "^3.9.1"
+        "@ui5/cli": "^3.6.0"
```
Comment (nikmace) [inline packages/adp-tooling/templates/projects/onprem/package.json:17]: Same question as before, is there any reason for this version "downgrade".

File: packages/adp-tooling/src/base/project-utils.ts [hunk 5]
```diff
@@ -0,0 +1,206 @@
...
+/**
+ * Writes a manifest.adpescr file within a specified folder in the project directory.
+ *
+ * @param {string} templatePath - The root path of the project template.
+ * @param {string} projectPath - The root path of the project.
+ * @param {ManifestAppdescr} data - The data to be populated in the template file.
+ * @param {Editor} fs - The `mem-fs-editor` instance used for file operations.
+ * @returns {void}
+ */
+export function writeManifestAppdecr(
```
Comment (nikmace) [inline packages/adp-tooling/src/base/project-utils.ts:49]: Spelling `writeManifestAppdecr` --> writeManifestAppde**s**cr

File: packages/adp-tooling/src/base/project-utils.ts [hunk 6]
```diff
@@ -0,0 +1,206 @@
...
+ * @param {Editor} fs - The `mem-fs-editor` instance used for file operations.
+ * @param {string[]} exclude - Files to be ignored when writing.
+ * @returns {void}
+ */
+export function writeTemplateToFolder(
+    templatePath: string,
+    projectPath: string,
+    data: CfModuleData | AdpWriterConfig,
+    fs: Editor,
+    exclude: string[] | [] = []
```
Comment (nikmace) [inline packages/adp-tooling/src/base/project-utils.ts:28]: Maybe as a suggestion for more clarity, this should be `ignoredFiles: string[] ...` or `excludedFiles`

**Review by nikmace — CHANGES_REQUESTED** [review]
GitHub actions for some reason are not running and I cannot see the result of those voters. Also, there are some merge conflicts to be resolved.

Some comments from before have not been addressed yet. And could you please write  what have you changed in my _requested changes comments_, so I can resolve those comments.

File: packages/adp-tooling/src/types.ts [hunk 7]
```diff
@@ -199,6 +227,29 @@ export const enum ChangeType {
...
+    CF = 'CF'
+}
+
+/**
+ * Maps a ChangeType to the corresponding data structure needed for that type of change.
+ * This conditional type ensures type safety by linking each change type with its relevant data model.
+ *
+ * @template T - A subtype of ChangeType indicating the specific type of change.
+ */
+export type ProjectGeneratorData<T extends ProjectType> = T extends ProjectType.ON_PREM
```
Comment (nikmace) [inline packages/adp-tooling/src/types.ts:245]: It would be good to remove the generator key word from the interface (if it is used by the create module there you are not using the generator)

File: packages/adp-tooling/src/types.ts [hunk 8]
```diff
@@ -347,3 +398,50 @@ export interface AdpProjectData {
     reference: string;
     id: string;
 }
+
+export interface AdpCustomConfig {
+    adp: {
+        safeMode: boolean;
+        environment: ProjectType;
```
Comment (nikmace) [inline packages/adp-tooling/src/types.ts:405]: TBD: the environment is now 'OnPrem' but our change editors still have checks for 'ABAP'

File: packages/adp-tooling/templates/projects/s4/package.json [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+    ],
+	"repository": {
+		"type": "git",
+		"build": "ui5.js build --verbose --include-task generateCachebusterInfo"
+	},
+	"author": "",
+	"license": "ISC",
+	"ui5": {
+		"dependencies": ["@sap/ui5-builder-webide-extension", "@ui5/task-adaptation"]
+	},
```
Comment (nikmace) [inline packages/adp-tooling/templates/projects/s4/package.json:20]: Looks like this file was not prettier'd and there are tabulations visible

File: packages/adp-tooling/templates/projects/cf/xs-security.json [hunk 10]
```diff
@@ -0,0 +1,7 @@
+{
+	"xsappname": "<%= projectName %>",
+	"tenant-mode": "dedicated",
+	"description": "Security profile of called application",
+	"scopes": [],
+	"role-templates": []
```
Comment (nikmace) [inline packages/adp-tooling/templates/projects/cf/xs-security.json:6]: This file is not prettier'd too

File: packages/adp-tooling/src/base/project-utils.ts [hunk 11]
```diff
@@ -0,0 +1,206 @@
...
+ */
+export function writeManifestAppdecr(
+    templatePath: string,
+    projectPath: string,
+    data: ManifestAppdescr,
+    fs: Editor
+): void {
+    try {
+        const appdescrTplPath = join(templatePath, 'webapp/manifest.appdescr_variant');
+        const appdescrPath = join(projectPath, 'webapp/manifest.appdescr_variant');
```
Comment (nikmace) [inline packages/adp-tooling/src/base/project-utils.ts:57]: Could you please use the enum from project access for different folders

File: packages/adp-tooling/src/base/project-utils.ts [hunk 12]
```diff
@@ -0,0 +1,206 @@
...
+ * Writes a .apd/config.json file within a specified folder in the project directory.
+ *
+ * @param {string} projectPath - The root path of the project.
+ * @param {CfAdpConfig} data - The data to be populated in the template file.
+ * @param {Editor} fs - The `mem-fs-editor` instance used for file operations.
+ * @returns {void}
+ */
+export function writeAdpConfig(projectPath: string, data: CfAdpConfig, fs: Editor): void {
+    try {
+        const adpConfigPath = join(projectPath, '.adp/config.json');
```
Comment (nikmace) [inline packages/adp-tooling/src/base/project-utils.ts:78]:
```suggestion
        const adpConfigPath = join(projectPath, '.adp', 'config.json');
```

File: packages/create/src/cli/generate/adaptation-project.ts [hunk 13]
```diff
@@ -78,7 +78,7 @@ async function generateAdaptationProject(
         if (!basePath) {
             basePath = join(process.cwd(), config.app.id);
         }
-        const fs = await generate(basePath, config);
+        const fs = await generate(basePath, ProjectType.ON_PREM, config);
```
Comment (nikmace) [inline packages/create/src/cli/generate/adaptation-project.ts:81]: I think this config is missing some properties, when you pass the project type to the method you might see ts errors.

```suggestion
        const fs = await generate<ProjectType.ON_PREM>(basePath, ProjectType.ON_PREM, config);
```

**Review by nikmace — CHANGES_REQUESTED** [review]

**Review by tobiasqueck — CHANGES_REQUESTED** [review]
Sorry but that one is far away from being acceptable.

Please remove things like a `pom.xml` that only make sense for SAP internal projects and also remove all the duplicated files in the templates section.
Projects that are for an ABAP system are almost identical (no matter if on premise, private cloud our public cloud). The only difference is the additional build task and a minor change in the deploy script.
I am not an expert for CF but I am pretty sure that the generated structure is also not aligned with the latest guidance on CF projects.

My recommendation is actually to start from scratch again. Take the existing writer/templates (from `main`) and compare its results with the three different options. Then enhance the "core" that is the same for all option and add diffs if required for other targets. I'd recommend you have a look at the `fiori-elements-writer` and how it is designed.

File: packages/adp-tooling/test/unit/writer/index.test.ts [hunk 14]
```diff
@@ -86,5 +86,146 @@ describe('ADP writer', () => {
...
+            const data = `ABAP_USERNAME: test_name
+                          ABAP_PASSWORD: test_pass`;
+            await generateEnv(projectDir, data, fs);
+            expect(
+                fs.dump(projectDir, (file) => file.dirname === projectDir && ['.env'].includes(file.basename))
+            ).toMatchSnapshot();
+        });
+    });
+
+    describe('generateCf', () => {
```
Comment (nikmace) [inline packages/adp-tooling/test/unit/writer/index.test.ts:155]: This can be removed as CF will not be part of this PR?

File: packages/adp-tooling/src/types.ts [hunk 15]
```diff
@@ -35,21 +35,41 @@ export interface AdpWriterConfig {
...
+        ui5Version?: string;
+        ui5EndpointUrl?: string;
     };
     package?: {
         name?: string;
         description?: string;
     };
+    flp?: {
+        flpTitle?: string;
+        flpSubtitle?: string;
```
Comment (nikmace) [inline packages/adp-tooling/src/types.ts:47]: Since these are under `flp` maybe we can omit this prefix `flp`? So it becomes this:
```suggestion
        title?: string;
        subtitle?: string;
```

File: packages/adp-tooling/src/types.ts [hunk 16]
```diff
@@ -35,21 +35,41 @@ export interface AdpWriterConfig {
     target: AbapTarget;
     ui5?: {
         minVersion?: string;
+        ui5Version?: string;
+        ui5EndpointUrl?: string;
```
Comment (nikmace) [inline packages/adp-tooling/src/types.ts:39]: Since these are under `ui5` maybe we can omit this prefix `ui5`? So it becomes this:
```suggestion
        version?: string;
        endpointUrl?: string;
```
Or just `url`?

**Review by nikmace — CHANGES_REQUESTED** [review]

File: packages/adp-tooling/test/unit/writer/index.test.ts (hunk 14 above)
Comment (IvoSG) [reply packages/adp-tooling/test/unit/writer/index.test.ts]: Done

File: packages/adp-tooling/src/types.ts (hunk 15 above)
Comment (IvoSG) [reply packages/adp-tooling/src/types.ts]: Done

File: packages/adp-tooling/src/types.ts (hunk 16 above)
Comment (IvoSG) [reply packages/adp-tooling/src/types.ts]: Done

File: packages/adp-tooling/templates/project/package.json [hunk 17]
```diff
@@ -10,17 +10,19 @@
...
+        <%if (locals.options?.fioriTools) {%>"@sap/ux-ui5-tooling": "1"<%} else {%>"@sap-ux/backend-proxy-middleware": "^0.7.5",
         "@sap-ux/preview-middleware": "^0.11.1",
         "@sap-ux/ui5-proxy-middleware": "^1.3.0",
         "@sap-ux/deploy-tooling": "^0.11.7"<%}%>,
-        "@ui5/cli": "^3.9.1"
+        "@ui5/task-adaptation": "^1.3.0",
+        "@ui5/cli": "^3.9.2"
     },
     "scripts": {
         "build": "ui5 build --exclude-task generateFlexChangesBundle generateComponentPreload --clean-dest",
```
Comment (tobiasqueck) [inline packages/adp-tooling/templates/project/package.json:21]: Could you please add `minify` to the excluded task? Otherwise, `.map` files will be generated that are rejected by LREP on your ABAP versions

File: packages/adp-tooling/src/writer/index.ts [hunk 18]
```diff
@@ -42,26 +49,18 @@ export async function generate(basePath: string, config: AdpWriterConfig, fs?: E
...
-        const ui5DeployConfig = await UI5Config.newInstance(baseUi5ConfigContent);
-        enhanceUI5DeployYaml(ui5DeployConfig, fullConfig);
-        fs.write(join(basePath, 'ui5-deploy.yaml'), ui5DeployConfig.toString());
+    writeTemplateToFolder(join(tmplPath, '**/*.*'), join(basePath), fullConfig, fs, ignoredFiles);
+    if (fullConfig.appdescr) {
+        writeManifestAppdescr(tmplPath, basePath, fullConfig.appdescr, fs);
     }
+    await writeUI5DeployYaml(basePath, fullConfig, fs);
+    await writeUI5Yaml(basePath, fullConfig, fs);
+    await writeEnvFile(basePath, fullConfig, fs);
```
Comment (tobiasqueck) [inline packages/adp-tooling/src/writer/index.ts:63]: Does the user know that we are writing user/password to the filesystem? If not, we cannot do that. 
Btw, as far as I understand, both user/password are only used for the additional tasks for S/4HANA Cloud Public Edition scenarios, however, such systems won't allow access with user/password, therefore, I'd strongly prefer to remove this user/password/env thing.

File: packages/adp-tooling/src/writer/options.ts [hunk 19]
```diff
@@ -120,3 +164,36 @@ function getOpenSourceMiddlewares(config: AdpWriterConfig): CustomMiddleware<obj
...
 }
+
+/**
+ * Get a list of required custom tasks for S4.
+ *
+ * @param config full project configuration
+ * @returns list of required tasks.
+ */
+function getAdpCloudCustomTasks(config: AdpWriterConfig & { target: AbapTarget }): CustomTask[] {
+    const user = 'env:ABAP_USERNAME';
```
Comment (tobiasqueck) [inline packages/adp-tooling/src/writer/options.ts:175]: As mentioned in my other comment, this shouldn't be here. In the very worst case that there is a valid scenario for this, use at least `FIORI_TOOLS_USER` and `FIORI_TOOLS_PASSWORD` because they are available if any previous Fiori tools task or middleware relied on user/password.

**Review by tobiasqueck — CHANGES_REQUESTED** [review]

File: packages/adp-tooling/src/types.ts [hunk 20]
```diff
@@ -35,21 +35,41 @@ export interface AdpWriterConfig {
...
     ui5?: {
         minVersion?: string;
+        version?: string;
+        url?: string;
     };
     package?: {
         name?: string;
         description?: string;
     };
+    flp?: {
```
Comment (tobiasqueck) [inline packages/adp-tooling/src/types.ts:45]: That object is odd. `bspName` and `credentials` are used to access the original app while `title`, `subtitle` and `inboundId` are only being used in the i18n file. Did you miss adding a change to the appdescr if these properties are present or did I miss it in the code or is there some magic backend functionality that does something on deploy? 

What is the difference between `bspName` and `reference` in the app section?

File: packages/adp-tooling/src/types.ts [hunk 21]
```diff
@@ -192,6 +213,14 @@ export const enum ChangeType {
     CHANGE_INBOUND = 'appdescr_app_changeInbound'
 }
 
+/**
+ * Enumerates the types of projects that can be made, each representing a specific kind of project structure.
+ */
+export const enum ProjectType {
```
Comment (tobiasqueck) [inline packages/adp-tooling/src/types.ts:219]: Both are S/4 systems here, so please change it to something like `Cloud` and `OnPremise`. Even better would be if you reuse the type `OperationsType` (https://github.com/SAP/open-ux-tools/blob/main/packages/axios-extension/src/abap/types/adt-types.ts#L77) because that is exactly what you will get back from the backend as identify during the prompting phase.

File: packages/adp-tooling/src/types.ts (hunk 8 above)
Comment (tobiasqueck) [reply packages/adp-tooling/src/types.ts:376]: @nikmace  Why would the change editors need to know about the target environment?

File: packages/adp-tooling/src/types.ts [hunk 22]
```diff
@@ -35,21 +35,41 @@ export interface AdpWriterConfig {
     target: AbapTarget;
     ui5?: {
```
Comment (tobiasqueck) [inline packages/adp-tooling/src/types.ts:36]: Could you please use `frameworkUrl` like https://github.com/SAP/open-ux-tools/blob/main/packages/ui5-application-writer/src/types.ts#L43, so that we keep both objects in sync.
Btw, what is the use case here? Is it for openUI5 or is there a different reason? It doesn't seem like an openUI5 use case since the target system is always an ABAP system, or did I miss something?

Comment (IvoSG) [reply packages/adp-tooling/src/types.ts:36]: It is about the internal scenario when the snapshot version is selected and points to an internal endpoint

File: packages/adp-tooling/src/types.ts (hunk 20 above)
Comment (IvoSG) [reply packages/adp-tooling/src/types.ts:45]: Currently, the title, subtitle and input id are only used in the i18n file because the appdescr information comes already filtered after the prompting.
bspName is the base app repoName  and reference is the base app id

File: packages/adp-tooling/src/writer/options.ts (hunk 19 above)
Comment (tobiasqueck) [reply packages/adp-tooling/src/writer/options.ts:175]: As discussed on the phone, please remove it.

File: packages/adp-tooling/src/writer/index.ts (hunk 18 above)
Comment (tobiasqueck) [reply packages/adp-tooling/src/writer/index.ts:63]: As discussed, please remove it.

File: packages/adp-tooling/src/types.ts (hunk 20 above)
Comment (tobiasqueck) [reply packages/adp-tooling/src/types.ts:45]: As discussed on the phone, please add the change for the inbound communication to the appdescr template file with the condtion that the flp object is provided

File: packages/adp-tooling/src/types.ts (hunk 22 above)
Comment (tobiasqueck) [reply packages/adp-tooling/src/types.ts:36]: Thanks.

File: packages/adp-tooling/src/types.ts [hunk 23]
```diff
@@ -35,21 +35,41 @@ export interface AdpWriterConfig {
...
      */
-    deploy?: Adp;
+    deploy?: Adp | BspApp;
     options?: {
         /**
          * Optional: if set to true then the generated project will be recognized by the SAP Fiori tools
          */
         fioriTools?: boolean;
     };
+    appdescr?: ManifestAppdescr;
```
Comment (tobiasqueck) [inline packages/adp-tooling/src/types.ts:67]: As discussed on the phone, please reduce this to a change array (i.e. the content section of the appdescr) so that the caller can push additional changes but cannot affect the rest of the file.

File: packages/adp-tooling/templates/project/package.json (hunk 17 above)
Comment (IvoSG) [reply packages/adp-tooling/templates/project/package.json]: Done

File: packages/adp-tooling/src/writer/index.ts (hunk 18 above)
Comment (IvoSG) [reply packages/adp-tooling/src/writer/index.ts]: Done

File: packages/adp-tooling/src/types.ts (hunk 20 above)
Comment (IvoSG) [reply packages/adp-tooling/src/types.ts:47]: The change for the inbound it's more complex to be in the aprdescr template and because of that it's passed with other changes in the content

File: packages/adp-tooling/src/types.ts (hunk 23 above)
Comment (IvoSG) [reply packages/adp-tooling/src/types.ts]: Done

File: packages/adp-tooling/src/types.ts (hunk 21 above)
Comment (IvoSG) [reply packages/adp-tooling/src/types.ts]: Done

File: packages/adp-tooling/src/types.ts (hunk 20 above)
Comment (tobiasqueck) [reply packages/adp-tooling/src/types.ts:47]: If it is too complex for ejs, implement it as a plain function and then add it to the list of changes. Most importantly move the logic here because you have a reference to it in your generated i18n.properties file

File: packages/adp-tooling/templates/project/webapp/manifest.appdescr_variant [hunk 24]
```diff
@@ -7,6 +7,9 @@
   "namespace": "apps/<%- app.reference %>/appVariants/<%- app.id %>/",
   "version": "0.1.0",
   "content": [
+<%if(app.content){-%><% for(let change of app.content){-%>
+    <%- JSON.stringify(change) %><%-","%>
```
Comment (tobiasqueck) [inline packages/adp-tooling/templates/project/webapp/manifest.appdescr_variant:11]: It might be nice to align the spacing of your JSON with the rest of the document e.g. JSON.stringify(change, undefined, 4)

File: packages/adp-tooling/src/writer/options.ts (hunk 19 above)
Comment (IvoSG) [reply packages/adp-tooling/src/writer/options.ts]: done

File: packages/adp-tooling/templates/project/webapp/manifest.appdescr_variant (hunk 24 above)
Comment (IvoSG) [reply packages/adp-tooling/templates/project/webapp/manifest.appdescr_variant]: done

File: packages/adp-tooling/src/types.ts [hunk 25]
```diff
@@ -340,3 +364,59 @@ export interface AdpProjectData {
...
+            subTitle: string;
+            signature: {
+                parameters: object | string;
+                additionalParameters: "allowed";
+            };
+        };
+    };
+}
+
+export const enum FlpConfigurationType {
```
Comment (tobiasqueck) [inline packages/adp-tooling/src/types.ts:420]: What is this enum of one value?

Comment (IvoSG) [reply packages/adp-tooling/src/types.ts:420]: Currently, we are supporting only ADD_NEW_TILE, but expecting to be more in the future

File: packages/adp-tooling/src/types.ts [hunk 26]
```diff
@@ -52,6 +58,23 @@ export interface AdpWriterConfig {
     };
 }
 
+export interface FlpConfiguration {
+    title: string;
+    bspName: string;
```
Comment (tobiasqueck) [inline packages/adp-tooling/src/types.ts:63]: Is the BSP name part of the FLP configuration? Or is it something else?

File: packages/adp-tooling/src/types.ts [hunk 27]
```diff
@@ -52,6 +58,23 @@ export interface AdpWriterConfig {
...
+    languages: Language[];
+    inboundId: string;
```
Comment (tobiasqueck) [inline packages/adp-tooling/src/types.ts:65]: The type indicates that `inboundId` is mandatory but looking at the code it is not.

File: packages/adp-tooling/src/types.ts (hunk 25 above)
Comment (tobiasqueck) [reply packages/adp-tooling/src/types.ts:420]: Could you give me a (few) examples? And will these also be called during the project generation?

File: packages/adp-tooling/src/types.ts [hunk 28]
```diff
@@ -31,19 +32,24 @@ export interface AdpWriterConfig {
         reference: string;
         layer?: UI5FlexLayer;
         title?: string;
+        content?: Content[];
```
Comment (tobiasqueck) [inline packages/adp-tooling/src/types.ts:35]: You are only using it if it is a cloud system (if I read the code correctly) which I think is a bit odd because the API does not indicate that.

File: packages/adp-tooling/src/types.ts [hunk 29]
```diff
@@ -52,6 +58,23 @@ export interface AdpWriterConfig {
     };
 }
 
+export interface FlpConfiguration {
```
Comment (tobiasqueck) [inline packages/adp-tooling/src/types.ts:61]: This type needs more explanation, please add a short comment to each property so that it is easier to understand what this config is for.

File: packages/adp-tooling/src/writer/index.ts [hunk 30]
```diff
@@ -59,14 +49,15 @@ export async function generate(basePath: string, config: AdpWriterConfig, fs?: E
...
-    const fullConfig = setDefaults(config);
 
-    fs.copyTpl(join(tmplPath, '**/*.*'), join(basePath), fullConfig, undefined, {
-        globOptions: { dot: true },
-        processDestinationPath: (filePath: string) => filePath.replace(/gitignore.tmpl/g, '.gitignore')
-    });
+    const fullConfig = setDefaults(config);
 
-    await writeUi5Yaml(basePath, fullConfig, fs);
+    if (fullConfig.customConfig?.adp.environment === "C" && fullConfig.flp) {
```
Comment (tobiasqueck) [inline packages/adp-tooling/src/writer/index.ts:55]: Is the FLP config not valid for on prem systems?

**Review by tobiasqueck — CHANGES_REQUESTED** [review]
I added a few inline questions.

You have a few places where you set defaults if not provided as input e.g. the inboundId. I found three places where you do it, therefore, I'd recommend you create something like a `mergeWithDefaults` function and call it right in the beginning. Additionally, you have properties that are required if another property is not provided, therefore, you should do a validation of the input right at the beginning instead of in multiple places of the code. Please check https://github.com/SAP/open-ux-tools/blob/main/packages/fiori-elements-writer/src/index.ts#L64-L66, it has the validation and then the merging cleanly in one place at the beginning. Afterwards you can rely on a valid configuation.

File: .changeset/famous-panthers-occur.md [hunk 31]
```diff
@@ -0,0 +1,7 @@
+---
+'@sap-ux/adp-tooling': minor
+'@sap-ux/preview-middleware': patch
```
Comment (tobiasqueck) [inline .changeset/famous-panthers-occur.md:3]: Please remove preview-middleware and axios-extension from the changeset.

File: packages/adp-tooling/test/unit/writer/index.test.ts [hunk 32]
```diff
@@ -40,7 +40,7 @@ describe('ADP writer', () => {
     describe('generate', () => {
         test('minimal config', async () => {
             const projectDir = join(outputDir, 'minimal');
-            await generate(projectDir, config, fs);
+            await generate(projectDir, config);
```
Comment (tobiasqueck) [inline packages/adp-tooling/test/unit/writer/index.test.ts:43]: What is the reason for removing the parameter `fs` here?

File: packages/adp-tooling/src/writer/index.ts (hunk 30 above)
Comment (IvoSG) [reply packages/adp-tooling/src/writer/index.ts:51]: Yes

File: packages/adp-tooling/src/types.ts (hunk 28 above)
Comment (IvoSG) [reply packages/adp-tooling/src/types.ts:35]: Content is used by both onPrem and Cloud if it's provided

File: packages/adp-tooling/src/types.ts (hunk 25 above)
Comment (IvoSG) [reply packages/adp-tooling/src/types.ts:420]: As far as I know, there is expected to be a "CHANGE TILE" as well

Comment (IvoSG) [reply packages/adp-tooling/src/types.ts]: Removed

File: packages/adp-tooling/src/types.ts (hunk 29 above)
Comment (IvoSG) [reply packages/adp-tooling/src/types.ts]: Done

File: .changeset/famous-panthers-occur.md (hunk 31 above)
Comment (IvoSG) [reply .changeset/famous-panthers-occur.md]: Done

File: packages/adp-tooling/src/writer/options.ts [hunk 33]
```diff
@@ -153,3 +164,151 @@ function getOpenSourceMiddlewares(config: AdpWriterConfig): CustomMiddleware<obj
...
+    const inboundChangeContent = addInboundId ? getInboundChangeContentWithNewInboundID(flpConfiguration as NewInboundNavigation, app.id) : getInboundChangeContentWithExistingInboundId(flpConfiguration as ChangeInboundNavigation, app.id);
+    if (inboundChangeContent) {
+        const firstFlpChange = {
+            changeType: addInboundId ? "appdescr_app_addNewInbound" : "appdescr_app_changeInbound",
+            content: inboundChangeContent,
+            texts: {
+                "i18n": "i18n/i18n.properties"
+            }
+        };
+        const secondFlpChange = {
```
Comment (tobiasqueck) [inline packages/adp-tooling/src/writer/options.ts:303]: Could we have a more descriptive name than first and second change .e.g addInboundChange and removeOtherInboundsChange

File: packages/adp-tooling/src/types.ts [hunk 34]
```diff
@@ -56,6 +69,32 @@ export interface AdpWriterConfig {
...
+export interface NewInboundNavigation {
+    /** Represent business entities and can bundle applications that reflect a specific scenario. */
+    semanticObject: string;
+    /** Operations which can be performed on a semantic object. */
+    action: string;
+    //** Defined instance of the semantic object (e.g. by specifying the employee ID). */
+    additionalParameters?: object;
+    title: string;
+    subTitle?: string;
+    inboundId?: string;
```
Comment (tobiasqueck) [inline packages/adp-tooling/src/types.ts:88]: `inboundId` shouldn't exit on a new navigation, correct?

File: packages/adp-tooling/test/unit/writer/index.test.ts (hunk 32 above)
Comment (tobiasqueck) [reply packages/adp-tooling/test/unit/writer/index.test.ts:43]: Good point but we use the `fs` instance here to be able to write the output to the filessystem for debugging (check the `afterAll` function), thereofre, please revert your change here

File: packages/adp-tooling/test/unit/writer/index.test.ts [hunk 35]
```diff
@@ -88,6 +88,70 @@ describe('ADP writer', () => {
...
+            const projectDir = join(outputDir, 'minimal');
+            await generateEnv(projectDir, '');
+            expect(
+                fs.dump(projectDir, (file) => file.dirname === projectDir && ['.env'].includes(file.basename))
+            ).toMatchSnapshot();
+        });
+
+        test('generate env file for s4hana cloud projects', async () => {
+            const projectDir = join(outputDir, 'env');
+            const data = `ABAP_USERNAME: test_name
```
Comment (tobiasqueck) [inline packages/adp-tooling/test/unit/writer/index.test.ts:148]: Didn't we remove user/password from the code? Please also remove it from the test

Comment (IvoSG) [reply packages/adp-tooling/test/unit/writer/index.test.ts]: done

File: packages/adp-tooling/src/writer/options.ts (hunk 33 above)
Comment (IvoSG) [reply packages/adp-tooling/src/writer/options.ts]: done

File: packages/adp-tooling/test/unit/writer/index.test.ts (hunk 32 above)
Comment (IvoSG) [reply packages/adp-tooling/test/unit/writer/index.test.ts]: done

File: packages/adp-tooling/src/types.ts (hunk 26 above)
Comment (IvoSG) [reply packages/adp-tooling/src/types.ts]: Moved to the CloudApp interface

File: packages/adp-tooling/src/types.ts (hunk 34 above)
Comment (IvoSG) [reply packages/adp-tooling/src/types.ts]: Moved in internal type

File: packages/adp-tooling/templates/project/webapp/manifest.appdescr_variant [hunk 36]
```diff
@@ -7,6 +7,9 @@
   "namespace": "apps/<%- app.reference %>/appVariants/<%- app.id %>/",
   "version": "0.1.0",
   "content": [
+<%if(app.content){-%><% for(let change of app.content){-%>
+    <%- JSON.stringify(change, undefined, 4) %><%-","%>
```
Comment (nikmace) [inline packages/adp-tooling/templates/project/webapp/manifest.appdescr_variant:11]: Are you sure it should be 4 instead of 2? Most prettier configs have space set to 2.

**Review by nikmace — DISMISSED** [review]
Code looks good
Coverage is good
Did not test manually

File: packages/adp-tooling/templates/project/webapp/manifest.appdescr_variant (hunk 36 above)
Comment (IvoSG) [reply packages/adp-tooling/templates/project/webapp/manifest.appdescr_variant:11]: yes, because it is an inner object

File: packages/adp-tooling/src/types.ts [hunk 37]
```diff
@@ -175,6 +229,7 @@ export type IWriterData<T extends ChangeType> = IWriter<GeneratorData<T>>;
  *
  * @template T - The specific type of data the writer will handle, determined by the associated ChangeType.
  */
+
```
Comment (testojs) [inline packages/adp-tooling/src/types.ts:232]: Please remove this extra empty line as it detaches the jsdoc from the interface.

**Review by testojs — DISMISSED** [review]
- the code looks good
- good test coverage
- not tested manually

**Review by tobiasqueck — CHANGES_REQUESTED** [review]
* changeset: the writer functionality was there before but you added support for cloud projects, therefore, please adjust the text
* to make use of the writer in our `create` a small enhancement is required (#1964), I'd recommend merging it into your branch and then merging it together into main so that everything stays consistent
* code is well structured and well documented
* test coverage is sufficient

Please consider the first two points, otherwise, I am happy to approve

File: packages/adp-tooling/src/types.ts (hunk 37 above)
Comment (nikmace) [reply packages/adp-tooling/src/types.ts]: Removed in: [4e27dbf](https://github.com/SAP/open-ux-tools/pull/1775/commits/4e27dbff0ba748cc271d6d99bb08a05d7657dd4a)

**Comment by nikmace** [comment]
@tobiasqueck @testojs Comments are addressed, would you please do a review again?

**Review by testojs — APPROVED** [review]
Re-approving

**Review by tobiasqueck — APPROVED** [review]
All requested changes have been applied. Thank you!
